xdsl/dialects/func.py

```diff
@@ -47,6 +56,105 @@ def verify_(self) -> None:
                 "input types"
             )
 
+    @classmethod
+    def parse(cls, parser: Parser) -> FuncOp:
+        # Parse visibility keyword if present
+        visibility = parser.tokenizer.next_token_of_pattern(
+            re.compile("public|nested|private")
+        )
+        if isinstance(visibility, Span):
+            visibility = visibility.text
+
+        # Parse function name
+        name = parser.parse_symbol_name().data
+
+        def parse_fun_input():
+            ret = parser.parse_optional_argument()
+            if ret is None:
+                ret = parser.parse_optional_type()
+            if ret is None:
+                parser.raise_error("Expected argument or type")
+            return ret
+
+        # Parse function arguments
+        args = parser.parse_comma_separated_list(
+            parser.Delimiter.PAREN,
+            parse_fun_input,
+        )
+
+        # Check consistency (They should be either all named or none)
+        if isa(args, list[parser.Argument]):
+            entry_args = args
+            input_types = [a.type for a in args]
+            assert isa(input_types, list[Attribute])
```

Comment:
is there a situation when this might fail?

Reply:
No, this is a polish point for the parse API: the method returning the asserted value has a flag, that promises the method would fail rather than returning a value failing that assert.
AFAIK this can't be properly typed by pyright, so maybe we should actually split the method.
This is my way of keeping the PR's scope reasonable, there's always something else to perfect!

Reply:
would a cast be more appropriate then? Seems wasteful to validate something we know to be true

Reply:
Happy to use a cast instead
Would you mind explaining the concrete benefit somewhere? Sound like something we want to document if we want to spread accross the codebase, and I'm just curious; how can a dynamic language like Python avoid a check here?

Reply:
As far as I know, the cast has no runtime cost, and the isa check is actually pretty expensive. So if we can guarantee that a property holds, then I'd go with a cast over an assert isa, potentially over an assert isinstance too